### PR TITLE
fix: loss masking with pad eos collision

### PR DESCRIPTION
**Pull request by akoumpa**
# What does this PR do ?

Please see https://github.com/NVIDIA-NeMo/Automodel/issues/1336 for the problem description. 

# Changelog

- Add specific line by line info of high level changes in this PR.

# Before your PR is "Ready for review"

**Pre checks**:

- [ ] Make sure you read and followed [Contributor guidelines](https://github.com/NVIDIA-NeMo/Automodel/blob/main/CONTRIBUTING.md)
- [ ] Did you write any new necessary tests?
- [ ] Did you add or update any necessary documentation?

If you haven't finished some of the above items you can still open "Draft" PR.

# Additional Information

- Related to # (issue)

### CODE REVIEW

File: nemo_automodel/components/datasets/llm/formatting_utils.py [hunk 1]
```diff
@@ -24,6 +26,47 @@
...
+    """
+    if pad_token_id != eos_token_id:
+        return sequence == pad_token_id
+
+    mask = torch.zeros(sequence.shape, dtype=torch.bool, device=sequence.device)
+    non_pad_positions = (sequence != pad_token_id).nonzero(as_tuple=True)[0]
+    if non_pad_positions.numel() > 0:
+        last_content_idx = non_pad_positions[-1].item()
+        # last_content_idx + 1 → real EOS (keep), last_content_idx + 2 → padding
+        mask[last_content_idx + 2 :] = True
```
Comment (ZhiyuLi-Nvidia) [inline nemo_automodel/components/datasets/llm/formatting_utils.py:63]: Does it hold true when pad_token_id and eos_token_id are different?

**Review by ZhiyuLi-Nvidia — APPROVED** [review]

File: nemo_automodel/components/datasets/llm/formatting_utils.py [hunk 2]
```diff
@@ -32,15 +75,37 @@ def _pad_to_seq_length(sample, pad_token_id, seq_length):
...
 
 
+_warned_add_pad_token = set()
+
+
 def _add_pad_token(tokenizer):
     """Add pad token to tokenizer if not present."""
     pad_token_id = None
-    if getattr(tokenizer, "pad_token_id", None) is None:
+    if getattr(tokenizer, "pad_token_id", None) is None and not "no_pad_id" in _warned_add_pad_token:
```
Comment (oliverholworthy) [inline nemo_automodel/components/datasets/llm/formatting_utils.py]: This `_warned_add_pad_token` set is module-level, but it also gates the tokenizer.pad_token_id = tokenizer.eos_token_id assignment. On a second call with a different tokenizer that also has pad_token_id=None, the fallback won't execute, leaving pad_token_id as None. Should we only guard the `logger.warning()`, not the assignment itself, it the intent here is to only log the warning once.

File: nemo_automodel/_transformers/tokenization/nemo_auto_tokenizer.py [hunk 3]
```diff
@@ -64,6 +98,24 @@ def from_pretrained(cls, pretrained_model_name_or_path, *args, add_bos_token=Tru
         tokenizer.__class__ = type(cls.__name__, (cls, base_tokenizer_cls), {})
         return tokenizer
 
+    def apply_chat_template(self, conversation, *args, **kwargs):
```
Comment (oliverholworthy) [inline nemo_automodel/_transformers/tokenization/nemo_auto_tokenizer.py]: What's the connection between this apply_chat_template method and the pad/eos fix? If unrelated, would it make sense in a separate PR?

File: nemo_automodel/components/datasets/llm/formatting_utils.py (hunk 1 above)
Comment (akoumpa) [reply nemo_automodel/components/datasets/llm/formatting_utils.py:63]: Yeah, I think so.

File: nemo_automodel/_transformers/tokenization/nemo_auto_tokenizer.py (hunk 3 above)
Comment (akoumpa) [reply nemo_automodel/_transformers/tokenization/nemo_auto_tokenizer.py]: Thanks @oliverholworthy , I've reverted this.

File: nemo_automodel/components/datasets/llm/formatting_utils.py (hunk 2 above)
Comment (akoumpa) [reply nemo_automodel/components/datasets/llm/formatting_utils.py]: Thanks @oliverholworthy , I've fixed the issue.
